enterprise/micronaut/src/org/netbeans/modules/micronaut/MicronautExpressionLanguageTokenId.java

```diff
@@ -0,0 +1,240 @@
+
+        private final AtomicBoolean canceled = new AtomicBoolean();
+
+        private MicronautExpressionLanguageEmbeddingProvider() {
+            super(JavaSource.Phase.ELEMENTS_RESOLVED);
+        }
+
+        @Override
+        public void run(Parser.Result result, SchedulerEvent event) {
+            final CompilationInfo ci = CompilationInfo.get(result);
+            if (ci != null && !canceled.get()) {
+                final List<Pair<Integer, Integer>> literalBounds = new ArrayList<>();
+                final SourcePositions sp = ci.getTrees().getSourcePositions();
+                new TreePathScanner<Void, List<Pair<Integer, Integer>>>() {
+                    @Override
+                    public Void visitLiteral(LiteralTree node, List<Pair<Integer, Integer>> p) {
+                        if (!canceled.get() && node.getValue() instanceof String) {
+                            TreePath path = this.getCurrentPath();
+                            if (path.getParentPath() != null && path.getParentPath().getLeaf().getKind() == Tree.Kind.ASSIGNMENT
+                                    && path.getParentPath().getParentPath() != null && path.getParentPath().getParentPath().getLeaf().getKind() == Tree.Kind.ANNOTATION) {
+                                p.add(Pair.of((int) sp.getStartPosition(path.getCompilationUnit(), node), (int) sp.getEndPosition(path.getCompilationUnit(), node)));
+                            }
+                        }
+                        return null;
+                    }
+                }.scan(ci.getCompilationUnit(), literalBounds);
+                if (!literalBounds.isEmpty()) {
+                    try {
+                        final Document doc = ci.getDocument();
+                        if (doc != null) {
+                            RequestProcessor.getDefault().post(() -> {
+                                Runnable runn = () -> {
+                                    final TokenHierarchy<Document> hierarchy = TokenHierarchy.get(doc);
+                                    final Language<?> java = Language.find("text/x-java");
+                                    if (java != null && micronautExpressionLanguage != null) {
+                                        final TokenSequence<?> ts = hierarchy.tokenSequence(java);
+                                        if (ts != null) {
+                                            for (Pair<Integer, Integer> bound : literalBounds) {
+                                                ts.move(bound.first());
+                                                while (ts.moveNext() && ts.offset() < bound.second()) {
+                                                    TokenSequence<?> embedded = ts.embedded();
+                                                    if (embedded != null) {
+                                                        ts.removeEmbedding(embedded.language());
```

Comment:
Discussed offline - as this will apply to all string values of all annotations, it possibly affects other syntaxes (in micronaut-unrelated annotations). Better avoid replacing foreign embeddings, if exists (i.e. SQL)

Reply:
Affects limited to Micronaut projects supporting Expression Language. Also - syntax coloring for Expression Language is avoided when foreign embeddings are detected.